contracts/anchor/src/state.rs

```diff
@@ -0,0 +1,320 @@
+    ) -> bool {
+        if root == [0u8; 32] {
+            return false;
+        }
+
+        let mut i = read_curr_neighbor_root_index(store, chain_id).unwrap_or_default();
+        for _ in 0..ROOT_HISTORY_SIZE {
+            if let Ok(r) = read_neighbor_roots(store, (chain_id, i)) {
+                if r == root {
+                    return true;
+                }
+
+                if i == 0 {
+                    i = ROOT_HISTORY_SIZE - 1;
+                } else {
+                    i -= 1;
+                }
+            }
+        }
+
+        false
+    }
+
+    pub fn is_valid_neighbor_roots(&self, roots: &[[u8; 32]], store: &dyn Storage) -> bool {
+        assert!(
+            roots.len() == self.max_edges as usize - 1,
+            "Incorrect roots length"
+        );
+        for (i, edge) in self
+            .chain_id_list
+            .iter()
+            .map(|c_id| read_edge(store, *c_id).unwrap_or_default())
+            .enumerate()
+        {
+            if !self.is_known_neighbor_root(edge.chain_id, roots[i], store) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+// Anchor struct
+// TODO: Anchor should have an ERC20 attached
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct Anchor {
+    pub chain_id: u64,
+    pub deposit_size: Uint256,
+    pub merkle_tree: MerkleTree,
+    pub linkable_tree: LinkableMerkleTree,
+    // used_nullifiers: Mapping<[u8; 32], bool>,
+    // poseidon: PoseidonRef,
+    // verifier: AnchorVerifierRef,
+}
+
+pub const ANCHOR: Item<Anchor> = Item::new("anchor");
+
+// Struct to save the use of "nullifiers"
+pub const NULLIFIERS: Map<Vec<u8>, bool> = Map::new("used_nullifers");
+
+// "Poseidon hasher"
+pub const POSEIDON: Item<Poseidon> = Item::new("poseidon");
+
+// "AnchorVerifier"
+pub const ANCHORVERIFIER: Item<AnchorVerifier> = Item::new("anchor_verifier");
+
+// MerkleTree
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct MerkleTree {
+    pub levels: u32,
+    pub current_root_index: u32,
+    pub next_index: u32,
+    // pub filled_subtrees: HashMap<u32, [u8; 32]>,
+    // pub roots: HashMap<u32, [u8; 32]>
+}
+
+// MerkleTree "filled_subtrees" Map
+pub const FILLEDSUBTREES: Map<String, [u8; 32]> = Map::new("filled_subtrees");
+
+pub fn save_subtree(store: &mut dyn Storage, k: u32, data: &[u8; 32]) -> StdResult<()> {
+    FILLEDSUBTREES.save(store, k.to_string(), data)
+}
+
+pub fn read_subtree(store: &dyn Storage, k: u32) -> StdResult<[u8; 32]> {
+    FILLEDSUBTREES.load(store, k.to_string())
+}
+
+// MerkleTree Roots Map
+pub const MERKLEROOTS: Map<String, [u8; 32]> = Map::new("merkle_roots");
+
+pub fn save_root(store: &mut dyn Storage, k: u32, data: &[u8; 32]) -> StdResult<()> {
+    MERKLEROOTS.save(store, k.to_string(), data)
+}
+
+pub fn read_root(store: &dyn Storage, k: u32) -> StdResult<[u8; 32]> {
+    MERKLEROOTS.load(store, k.to_string())
+}
+
+impl MerkleTree {
```

Comment:
Can we reuse the implementations of merkle trees between mixer/anchor?

Reply:
I think we cannot reuse the implementations of merkle trees between mixer/anchor.
The reason is that merkletree implementation depends on state/storage mutate functions, like "save_root" and "save_subtree".

Reply:
Right, makes sense.